idaes/core/util/utility_minimization.py

```diff
@@ -0,0 +1,621 @@
+##############################################################################
+"""
+General purpose heat integration block for IDAES models
+"""
+
+# Import Python libraries
+import logging
+
+# Import Plotting and Numpy libraries for composite curves
+import matplotlib.pyplot as plt
+import numpy as np
+
+# Import Pyomo libraries
+from pyomo.environ import (Constraint,
+                           sqrt,
+                           Expression,
+                           value,
+                           Var)
+
+from pyomo.environ import units as pyunits
+
+__author__ = "Alejandro Garciadiego, Alexander Dowling"
+
+def min_utility(blk, heating, cooling, DTmin, eps=1e-6, DG_units=pyunits.Mwatt):
+    """
+    Function for Duran-Grossmann for minimization of utilities
+    This function adds variables and constraints to the flowsheet to minimize utilities
+    Allows the calculation of reactors
+
+    Args:
+        blk: flowsheet
+        heating: Equipment that requires Heat
+        cooling: Equipment from wich heat is being removed
+        DTmin: HRAT (Heat Recovery Approximation Temperature)
+        eps: Epsilon for smoothing operation
+
+    Returns:
+        Constraint and variable object representing the calculation for hot utility and cold utility
+    """
+    epsT = 1
+
+    # Generate lists out of strings from Args
+    exchanger_list = heating + cooling
+    pinch_streams = exchanger_list
+
+    # Generate dictionaries out of strings to convert to pyomo objects
+    pinch_streamsdict = {}
+    coolingdict = {}
+    heatingdict = {}
+    exchangerdict = {}
+
+    for i,el in enumerate(exchanger_list):
+        pinch_streamsdict[str(el)] = el
+
+    for i,el in enumerate(cooling):
+        coolingdict[str(el)] = el
+
+    for i,el in enumerate(heating):
+        heatingdict[str(el)] = el
+
+    for i,el in enumerate(exchanger_list):
+        exchangerdict[str(el)] = el
+
+    Q = {}
+
+    for i in exchangerdict:
+        Q[i] = pyunits.convert(pinch_streamsdict[i].heat_duty[0],
+                       to_units=DG_units)
+
+    # Run function to fill exchanger data for pinch calculations for initialization of variables
+    exchangeData = heat_data(blk, heating, cooling, DG_units)
+
+    # Call pinch calculation for initialization of variables in PD class
+    PD = pinch_calc(heating, cooling, exchangeData, DTmin, eps)
+
+    # Define dictionary for addition of DTmin to heating equipment
+    dT = {}
+
+    for i,el in enumerate(cooling):
+        dT[str(el)] = 0
+
+    for i,el in enumerate(heating):
+        dT[str(el)] = DTmin
+
+    def T_in(blk, i):
+        return pinch_streamsdict[i].control_volume.properties_in[0].temperature
+
+    blk.Tin = Expression(pinch_streamsdict.keys(),
+                    rule=T_in,
+                    doc='Inlet temperature in exchangers [K]')
```

Comment:
It would be good to add Pyomo units here. However, this seems to presume that SI units are being used.

Reply:
Delete [K] from doc, so we do not assume what temperature we should use.